sympy/polys/matrices/domainmatrix.py

```diff
@@ -770,6 +775,31 @@ def matmul(A, B):
         A._check('*', B, A.shape[1], B.shape[0])
         return A.from_rep(A.rep.matmul(B.rep))
 
+    def scalarmul(A, lamda):
+        if lamda.element == lamda.domain.zero:
+            m, n = A.shape
+            return DomainMatrix([[lamda.domain.zero]*n]*m, (m, n), A.domain)
+        if lamda.element == lamda.domain.one:
+            return A
+
+        return A.mul(lamda.element)
+
+    def __truediv__(A, lamda):
+        """ Method for Scalar Divison"""
+        if isinstance(lamda, int):
+            lamda = DomainScalar(ZZ(lamda), ZZ)
+
+        if not isinstance(lamda, DomainScalar):
+            return NotImplemented
+
+        A, lamda = A.unify(lamda)
+        if lamda.element == lamda.domain.zero:
+            raise ZeroDivisionError
+        if lamda.element == lamda.domain.one:
+            return A
+
+        return A.__mul__(DomainScalar.from_sympy(1/lamda.element))
```

Comment:
This should not use `from_sympy`.

Also this should check whether the domain is a field or otherwise convert to a field before using `1/lambda.element`.

Reply:
> This should not use `from_sympy`.
>
We are using `DomainScalar.from_sympy` inorder to construct domain for `1/lamda.element`. 
I'm not aware if there is some other way in order to find its domain.
 
> Also this should check whether the domain is a field or otherwise convert to a field before using `1/lambda.element`.
>
Since `DomainScalar.from_sympy(1/lamda.element)` will return a `DomainScalar`, further, we are calling `DomainMatrix.__mul__`, we don't need to worry about any division as we will be unifying the domain in it.  

Kindly share your opinion.
Thanks.

Reply:
The domain for `lamda` is `lamda.domain`. It isn't possible to divide in all domains though e.g. `ZZ` so probably this should convert to a field like:
```python
A, lamda = A.to_field().unify(lamda)
```

Reply:
> The domain for `lamda` is `lamda.domain`. 
I mean to ask for `1/lamda.element`.

>It isn't possible to divide in all domains though e.g. `ZZ` so probably this should convert to a field like:
> 
This is the reason why we took the element as `1/lamda.element` and construct the DomainScalar using `from_sympy`.

@oscarbenjamin, I'm not able to understand your point. 
Can you please explain with some example or edge case I am missing to understand?
Thanks a lot.

Reply:
The point is that dividing a `ZZ` matrix by a `ZZ` scalar should not give an `RR` matrix. The result could be `ZZ` if the division can be computed exactly in `ZZ`. Or it could be `QQ`. Probably it's best just to make it always go to `QQ` or otherwise to raise an error when the domain is not a field. In that case the caller would be responsible for converting to a field domain.

If you want to to convert the domain of the matrix into one in which true division is possible then `DomainMatrix` has a `to_field` method for doing exactly that.

To see an example in which this fails with `TypeError` try dividing by an element of `ZZ[x]`. It is not generally possible to divide elements of non-field rings.

Reply:
Sure, thanks a lot for the explanation, it is clear to me now.